lib/src/dartpad_support_server.dart

```
@@ -56,7 +52,63 @@ class FileRelayServer {
+  Future<UuidContainer> getValidId() async {
+    int count = 0;
+    int limit = 4;
+    var database = ae.context.services.db;
+    String randomUuid;
+    var query;
+    List result;
+    do {
+      randomUuid = new uuid_tools.Uuid().v4();
+      query = database.query(_GistMapping)..filter('internalId =', randomUuid);
+      result = await query.run().toList();
+      count ++;
+    } while (!result.isEmpty && count < limit);
+    if (!result.isEmpty) {
+      _logger.severe("Could not generate valid ID.");
+      return new Future.value(new UuidContainer());
+    }
+    _logger.info("Valid ID ${randomUuid} retrieved.");
+    return new Future.value(new UuidContainer.fromUuid(randomUuid));
+  }
+
+  @ApiMethod(method: 'POST', path: 'storeGist')
+  Future<UuidContainer> storeGist(Mapping map) async {
+    var database = ae.context.services.db;
+    var query = database.query(_GistMapping)..filter('internalId =', map.internalId);
+    List result = await query.run().toList();
+    if (!result.isEmpty) {
+      _logger.severe("Collision with mapping of Id ${map.gistId}.");
+      return new Future.value(new UuidContainer());
+    } else {
+      _GistMapping entry = new _GistMapping.fromMap(map);
+      db.dbService.commit(inserts: [entry]).catchError((e) {
+        _logger.severe("Error while recording mapping with Id ${map.gistId}. Error ${e}");
+        throw e;
+      });
+      _logger.info("Mapping with ID ${map.gistId} stored.");
+      return new Future.value(new UuidContainer.fromUuid(map.gistId));
+    }
+  }
+
+  @ApiMethod(method: 'POST', path: 'retrieveGist')
+  Future<UuidContainer> retrieveGist(UuidContainer id) async {
+    var database = ae.context.services.db;
+    var query = database.query(_GistMapping)..filter('internalId =', id);
+    List result = await query.run().toList();
+    if (result.isEmpty) {
+      _logger.severe("Missing mapping for Id ${id.uuid}.");
+      return new Future.value(new UuidContainer.fromUuid(""));
```

This should return an Error 400